api/bucket.go

```diff
@@ -1,19 +1,205 @@
 package api
 
 import (
-	"github.com/gin-gonic/gin"
-	"github.com/konveyor/tackle2-hub/model"
+	"archive/tar"
+	"bufio"
+	"bytes"
+	"compress/gzip"
+	"fmt"
 	"io"
+	"io/ioutil"
 	"net/http"
 	"os"
+	"path"
 	pathlib "path"
+	"path/filepath"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	"github.com/konveyor/tackle2-hub/model"
 )
 
 //
 // BucketHandler provides bucket management.
 type BucketHandler struct {
 }
 
+// TODO: Validate Wildcard param to not allow access dirs outside of the bucket
+func (h *BucketHandler) serveBucketGet(ctx *gin.Context, owner *model.BucketOwner) {
+	if ctx.Request.Header.Get(Accept) == TarGzMimetype {
+		h.getDirArchive(ctx, path.Join(owner.Bucket, ctx.Param(Wildcard)))
+	} else {
+		h.content(ctx, owner)
+	}
+}
+
+func (h *BucketHandler) serveBucketUpload(ctx *gin.Context, owner *model.BucketOwner) {
+	if ctx.Request.Method == "PUT" {
+		h.uploadDirArchive(ctx, path.Join(owner.Bucket, ctx.Param(Wildcard)))
+	} else {
+		h.upload(ctx, owner)
+	}
+}
+
```

Comment:
The uploadDirArchive method takes a gin ctx with an uploaded ```file``` pointing to the tar.gz archive and a string specifying a destination directory where the archive should be extracted to. The directory needs to exist before and extration replaces all previous content.